app/scripts/lib/nonce-tracker.js

```diff
@@ -83,12 +76,80 @@ class NonceTracker {
     return mutex
   }
 
+  async _getNetworkNonceAndDetails (address) {
+    // calculate next nonce
+    // we need to make sure our base count
+    // and pending count are from the same block
+    const currentBlock = await this._getCurrentBlock()
+    const blockNumber = currentBlock.blockNumber
+    const pendingNonce = this._getLocalPendingNonce(address)
+    const pendingCount = this._getPendingTransactionCount(address)
+    assert(Number.isInteger(pendingCount), `nonce-tracker - pendingCount is not an integer - got: (${typeof pendingCount}) "${pendingCount}"`)
+    const baseCountHex = await this._getTxCount(address, currentBlock)
+    const baseCount = parseInt(baseCountHex, 16)
+    assert(Number.isInteger(baseCount), `nonce-tracker - baseCount is not an integer - got: (${typeof baseCount}) "${baseCount}"`)
+    // if the nonce provided by the network is higher then a pending tx
+    // toss out the pending txCount
+    const networkNonce = pendingNonce > baseCount ? baseCount + pendingCount : baseCount
+
+    return {networkNonce, blockNumber, baseCountHex, baseCount, pendingCount, pendingNonce}
+  }
+
+  _getLocalNextNonce (address) {
+    const confirmedTransactions = this._reduceTxListToUniqueNonces(this.getConfirmedTransactions(address))
+    const pendingTransactions = this._reduceTxListToUniqueNonces(this.getPendingTransactions(address))
+    const transactions = this._reduceTxListToUniqueNonces(confirmedTransactions.concat(pendingTransactions))
+    let localNonce = this._getHighestNonce(transactions)
+    // throw out localNonce if not a number
+    if (!Number.isInteger(localNonce)) localNonce = 0
+    if (
+      // the local nonce is not 0
+      localNonce ||
+      // or their are pending or confirmed transactions
+      this._getPendingTransactionCount(address) ||
+      confirmedTransactions.length
+    ) ++localNonce
+    return localNonce
+  }
+
+  _getLocalPendingNonce (address) {
+    const pendingTransactions = this._reduceTxListToUniqueNonces(this.getPendingTransactions(address))
+    const localNonce = this._getHighestNonce(pendingTransactions)
+    return localNonce
+  }
+
+  _getPendingTransactionCount (address) {
+    const pendingTransactions = this.getPendingTransactions(address)
+    return this._reduceTxListToUniqueNonces(pendingTransactions).length
```

Comment:
I was surprised to see a `.length` got past the new tests. I don't think counting pending transactions is as useful as looking at their last nonce. To demonstrate this problem, I've written a new test that this fails. Let me know if you think it makes sense to enforce.